src/broker.toit

```diff
@@ -0,0 +1,390 @@
+          throw "INVALID_SUBSCRIPTION: $topic"
+
+      if not 0 <= topic_qos.max_qos <= 2:
+        throw "INVALID_SUBSCRIPTION: $topic ($topic_qos.max_qos)"
+
+      accepted_qos := min topic_qos.max_qos 1
+      subscription_tree_.set topic accepted_qos
+      result_qos.add accepted_qos
+    send_ (SubAckPacket --qos=result_qos --packet_id=packet.packet_id)
+
+    packet.topics.do: | topic_qos/TopicQos |
+      topic := topic_qos.topic
+      broker.retained.do topic --all: | retained/PublishPacket |
+        qos := min topic_qos.max_qos  retained.qos
+        packet_id := qos > 0 ? next_packet_id_++ : null
+        send_ (retained.with --packet_id=packet_id --retain --qos=qos)
+
+  unsubscribe packet/UnsubscribePacket:
+    packet.topics.do: | topic |
+      existed := subscription_tree_.remove topic
+      if not existed:
+        logger_.info "client $client_id unsubscribed from non-existent topic $topic"
+    send_ (UnsubAckPacket --packet_id=packet.packet_id)
+
+  publish packet/PublishPacket:
+    topic := packet.topic
+    if topic == "" or topic.contains "#" or topic.contains "+":
+      throw "INVALID PUBLISH TOPIC. NO WILD CARDS ALLOWED. $packet.topic"
+    needs_ack := packet.qos > 0
+    if needs_ack:
+      packet_id := packet.packet_id
+      send_ack_ (PubAckPacket --packet_id=packet_id)
+
+    broker.publish packet
+
+  ping:
+    send_ (PingRespPacket)
+
+  disconnect --reason=null -> none:
+    if state_ == STATE_DISCONNECTED_: return
+    state_ = STATE_DISCONNECTED_
+    reason_suffix := reason ? " ($reason)" :""
+    logger_.info "client $client_id closing$reason_suffix"
+    if connection_:
+      connection_.close
+      connection_ = null
+
+    // Send the last will before we kill all tasks.
+    // Otherwise we will cancel the task on which we currently run on.
+    if reason and last_will_:
+      packet_id := last_will_.qos > 0 ? next_packet_id_++ : null
+      packet := PublishPacket last_will_.topic last_will_.payload \
+          --qos=last_will_.qos --packet_id=packet_id --retain=last_will_.retain
+      broker.publish packet
+
+    if clean_session: broker.remove_session_ client_id
+
+    if reader_task_:
+      reader_task_.cancel
+      reader_task_ = null
+    if writer_task_:
+      writer_task_.cancel
+      writer_task_ = null
+
+
+  send_ packet/Packet:
+    queued_.add_packet packet
+
+  send_ack_ ack/Packet:
+    queued_.add_ack ack
+
+  dispatch_incoming_publish packet/PublishPacket:
+    // There doesn't seem to be a rule which qos we should use if multiple
+    // subscriptions match. We thus use the one from the most specialized.
+    subscription_tree_.do --most_specialized packet.topic: | subscription_max_qos |
+      qos := min packet.qos subscription_max_qos
+      NO_PACKET_ID ::= -1  // See $PublishPacket.with.
+      packet_id := qos > 0 ? next_packet_id_++ : NO_PACKET_ID
+      send_ (packet.with --packet_id=packet_id --qos=qos)
+
+
+/** An unbounded channel for publish messages. */
```

Comment:
This could also be a semaphore and a deque.

Reply:
done.